system/datatypes/mailinglist.php

```diff
@@ -0,0 +1,728 @@
+				$hyphaDictionary['ml_email'] = 'email';
+				$hyphaDictionary['ml_subscribe'] = 'subscribe';
+				$hyphaDictionary['ml_successfully_updated'] = 'Successfully updated';
+				$hyphaDictionary['ml_successfully_subscribed'] = 'Successfully subscribed';
+				$hyphaDictionary['ml_successfully_unsubscribed'] = 'Successfully unsubscribed';
+				$hyphaDictionary['ml_successfully_created'] = 'Successfully created';
+				$hyphaDictionary['ml_successfully_sent'] = 'Successfully sent';
+				$hyphaDictionary['ml_successfully_submitted_email'] = 'Successfully sent, please confirm the email';
+				$hyphaDictionary['ml_status'] = 'status';
+				$hyphaDictionary['ml_draft'] = 'draft';
+				$hyphaDictionary['ml_sent'] = 'sent';
+				$hyphaDictionary['ml_send'] = 'send';
+				$hyphaDictionary['ml_addresses'] = 'addresses';
+				break;
+		}
+	}
+
+	public function indexAction() {
+		// check if form is posted and get form data
+		$formData = [];
+		$formPosted = $this->isPosted(self::FORM_CMD_LIST_SUBSCRIBE);
+		if ($formPosted) {
+			$formData = $_POST;
+		}
+
+		// create form
+		$form = $this->createSubscribeForm($formData);
+
+		// process form if it was posted
+		if ($formPosted) {
+			$form->validateRequiredField(self::FIELD_NAME_EMAIL);
+			$form->validateEmailField(self::FIELD_NAME_EMAIL);
+			if (empty($form->errors)) {
+				// check if email is already in the list, no need to have it twice.
+				$email = strtolower($form->dataFor(self::FIELD_NAME_EMAIL));
+				$addresses = $this->getDoc()->getOrCreate('addresses');
+				$address = $addresses->children()->findXPath('//address[@email="' . $email . '"]')->first();
+
+				if (!$address instanceof HyphaDomElement) {
+					$addAddress = true;
+					$sendEmail = true;
+					$code = uniqid();
+				} else {
+					// if the email is there just send another email if it is still pending
+					$sendEmail = $address->getAttribute('status') == self::ADDRESS_STATUS_PENDING;
+					$addAddress = false;
+					$code = $address->getAttribute('code');
+				}
+
+				if ($addAddress) {
+					// add email with status pending
+					$this->xml->lockAndReload();
+					$address = $this->xml->createElement('address');
+					$address->generateId();
+					$address->setAttribute('email', $email);
+					$address->setAttribute('status', self::ADDRESS_STATUS_PENDING);
+					$address->setAttribute('confirm-code', $code);
+					$this->getDoc()->getOrCreate('addresses')->append($address);
+					$this->xml->saveAndUnlock();
+				}
+
+				if ($sendEmail) {
+					// send email so that action can be confirmed
+					$link = $this->generateFullPath(sprintf('%s/confirm?email=%s&code=%s', $this->pagename, $email, $code));
+					$subject = hypha_getTitle() . ' - ' . __('ml_confirm_email_subject');
+					$this->send($subject, '<a href="' . $link . '">Please confirm your email address</a>', [$email]);
+
+					notify('success', __('ml_successfully_submitted_email'));
+				} else {
+					notify('success', __('ml_successfully_subscribed'));
```

Comment:
I wonder if here (and some other place below) it might be better to add an "already subscribed" (or below, "already confirmed") message, to give the user a bit more feedback about what is (not) happening.

Reply:
The current behaviour is based on a design principle (I forgot the name) that the output should always be consistent. Also the receiving party does not care that it is 'already subscribed', it just cares that it got 'subscribed' (of confirmed or unsubscribed). We can talk about it if you want.

Reply:
Your argument would then be that the output is only based on the resulting state, not on the actions performed. It relates a bit to the concept of idempotency. I have no strong preference in this.